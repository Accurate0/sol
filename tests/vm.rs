use insta::assert_compact_debug_snapshot;
use plrs::{
    compiler::Compiler,
    lexer::Lexer,
    parser::Parser,
    types,
    vm::{VMValue, VM},
};

#[test]
fn complex_math() {
    let input = r#"
        let z = (2 * 2) / ((3 - 4) * 2);
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn math() {
    let input = r#"
        let x = 1 + 2 / 3;
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn prefix() {
    let input = r#"
        let x = -1;
        let y = -(x + 3);
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn prefix_boolean() {
    let input = r#"
        let x = true;
        let y = !x;
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn native_function() {
    let input = r#"
        test_function();
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program).define_native_function("test_function".to_owned(), |_| None);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn native_function_with_return_value() {
    let input = r#"
        let x = test();
        if x {
            print("pass");
        } else {
            print("fail");
        }
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program).define_native_function("test".to_owned(), |_| {
        Some(VMValue::Literal(std::borrow::Cow::Owned(
            types::Literal::Boolean(true),
        )))
    });
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn nested_loop() {
    let input = r#"
let mut x = 0;
loop {
    let mut y = 0;
    loop {
        if y > 3 {
            print("exit loop");
            break;
        }

        y = y + 1;
        print(y);
    }

    if x > 3 {
        print("exit loop");
        break;
    }

    x = x + 1;
    print(x);
}
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}

#[test]
fn objects() {
    let input = r#"
let y = 3;

let another_object = {
    inner_value: 32,
};

let x = {
    test: 1,
    test2: "testing",
    test3: y,
    test4: another_object,
    test5: {
        test6: {
            test7: 1999
        }
    }
};

print(x);
print(x.test);
print(x.test2);
print(x.test3);
print(x.test4);
print(x.test4.inner_value);
print(x.test5);
print(x.test5.test6);
print(x.test5.test6.test7);
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let program = compiler.compile().unwrap();

    let vm = VM::new(program);
    let register_state = vm.run_with_registers_returned();

    assert_compact_debug_snapshot!(register_state);
}
