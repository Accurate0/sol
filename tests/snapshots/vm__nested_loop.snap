---
source: tests/vm.rs
expression: register_state
---
Ok([Empty, Literal(Integer(4)), Literal(Integer(4)), Literal(Integer(3)), Literal(Boolean(true)), Literal(String("exit loop")), Literal(String("exit loop")), Literal(String("print")), Empty, Literal(Integer(1)), Literal(Integer(4)), Literal(Integer(4)), Literal(String("print")), Empty, Literal(Integer(3)), Literal(Boolean(true)), Literal(String("exit loop")), Literal(String("exit loop")), Literal(String("print")), Empty, Literal(Integer(1)), Literal(Integer(4)), Literal(Integer(4)), Literal(String("print")), Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty])
